### DISCUSSION TITLE
Supporting real union embedding without pointers in Sum Types

### DISCUSSION THREAD

**Discussion by bbodi** [vlang/v#11807] [Ideas]
Hi, first of all, really great project!

Currently Sum Types behavior can be misleading to the user and problematic from performance perspective. The following example from the documentations allocates on the heap, while it seems like a normal stack allocation.
```
struct Moon {}

struct Mars {}

struct Venus {}

type World = Mars | Moon | Venus

_ := Moon{} // stack alloc
sum := World(Moon{}) // heap alloc
assert sum.type_name() == 'Moon'
println(sum)
```

c def of World:
```
// Union sum type main__World = 
//          |   75 = main__Mars          
//          |   74 = main__Moon          
//          |   76 = main__Venus         
struct main__World {
    union {
        main__Mars* _main__Mars;
        main__Moon* _main__Moon;
        main__Venus* _main__Venus;
    };
    int typ;
};
```

I would like to be able to iterate through an array of Sum Type, and dereferencing on each element can result in poor  performance.

Do you consider supporting real union embedding without pointers in Sum Types?

**Comment by bbodi** [comment]
Hey guys,

Any update on this?

**Comment by islonely** [comment]
I'd like to see this. I have had similar issues with sum types in the past.

A temporary solution would just be to make your own struct. V has unions but you have to access them within unsafe blocks.
